core/src/main/scala/com/salesforce/op/utils/stages/NameDetectUtils.scala

```diff
@@ -0,0 +1,337 @@
+
+import scala.io.Source
+import scala.util.Try
+
+/**
+ * Provides shared helper functions and variables for name identification and name to gender transformation.
+ */
+private[op] trait NameDetectFun extends Logging with NameDetectParams {
+  import GenderDetectStrategy._
+  import NameDetectUtils._
+
+  private[op] implicit def textValue2String[T <: Text](input: T#Value): String = input.getOrElse("")
+  private[op] implicit def string2Text[T <: Text](input: String): T#Value = input match {
+    case "" => Text.empty.value
+    case _ => Text(input).value
+  }
+
+  private[op] def preProcess(input: String): Seq[String] = {
+    TextTokenizer.tokenize(Text(input)).tokens.toArray
+  }
+
+  private[op] def computeGuardCheckQuantities(
+    text: String,
+    tokens: Seq[String],
+    hllMonoid: HyperLogLogMonoid
+  ): GuardCheckStats = {
+    val textLength = text.length
+    GuardCheckStats(
+      countBelowMaxNumTokens = if (tokens.length < 10) 1 else 0,
+      countAboveMinCharLength = if (textLength > 2) 1 else 0,
+      approxMomentsOfTextLength = Moments(textLength),
+      approxNumUnique = hllMonoid.create(text.getBytes)
+    )
+  }
+
+  private[op] def performGuardChecks(stats: GuardCheckStats, hllMonoid: HyperLogLogMonoid): Boolean = {
+    val N: Double = stats.approxMomentsOfTextLength.count.toDouble
+    val checks = List(
+      // check that in at least 3/4 of the texts there are no more than 10 tokens
+      (stats.countBelowMaxNumTokens / N) > 0.75,
+      // check that at least 3/4 of the texts are longer than 3 characters
+      (stats.countAboveMinCharLength / N) > 0.75,
+      // check that the standard deviation of the text length is greater than a small number
+      N < 10 || stats.approxMomentsOfTextLength.stddev > 0.05,
+      // check that the number of unique entries is at least 10
+      N < 100 || hllMonoid.sizeOf(stats.approxNumUnique).estimate > 10
+    )
+    checks.forall(identity)
+  }
+
+  private[op] def dictCheck(tokens: Seq[String], dict: NameDictionary): Double = {
+    if (tokens.isEmpty) 0.0 else {
+      tokens.map({ token: String => if (dict contains token) 1 else 0}).sum.toDouble / tokens.length
+    }
+  }
+
+  private[op] def genderDictCheck(nameToCheckGenderOf: String, genderDict: GenderDictionary): String = {
+    genderDict.get(nameToCheckGenderOf).map(
+      probMale => if (probMale >= 0.5) Male else Female
+    ).getOrElse(GenderNA)
+  }
+
+  private[op] def identifyGender(
+    text: String,
+    tokens: Seq[String],
+    strategy: GenderDetectStrategy,
+    genderDict: GenderDictionary
+  ): String = {
+    strategy match {
+      case FindHonorific() =>
+        val matched = tokens filter { NameDetectUtils.AllHonorifics contains }
+        if (matched.length == 1) {
+          if (MaleHonorifics contains matched.head) Male else Female
+        }
+        else GenderNA
+      case ByIndex(index) =>
+        val nameToCheckGenderOf = tokens.lift(index).getOrElse("")
```

Comment:
again, I think it's a bad idea to operate on `""`, instead either use `Option[String]` or `Text.empty`